src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,229 @@
+#[allow(missing_docs)]
+pub struct BreadcrumbsModel {
+    logger         : Logger,
+    animations     : Animations,
+    display_object : display::object::Instance,
+    scene          : Scene,
+    breadcrumbs    : Rc<RefCell<Vec<Breadcrumb>>>
+}
+
+impl BreadcrumbsModel {
+    /// Create new ProjectNameModel.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S,frp:&Frp) -> Self {
+        let scene          = scene.into();
+        let logger         = Logger::new("Breadcrumbs");
+        let display_object = display::object::Instance::new(&logger);
+        let animations     = Animations::new(&frp.network);
+        let scene          = scene.clone_ref();
+        let breadcrumbs    = Rc::new(RefCell::new(default()));
+        Self{logger,display_object,animations,scene,breadcrumbs}
+    }
+
+    fn width(&self) -> f32 {
+        let mut width = 0.0;
+        for breadcrumb in self.breadcrumbs.borrow().iter() {
+            width += breadcrumb.width();
+        }
+        width
+    }
+
+    fn select_breadcrumb(&self, index:usize) {
+        // If we have more crumbs after `index`, we will remove them.
+        let split_index = index + 1;
+        if split_index < self.breadcrumbs.borrow().len() {
+            let removed_breadcrumbs = self.breadcrumbs.borrow_mut().split_off(split_index);
+            for removed_breadcrumb in removed_breadcrumbs {
+                removed_breadcrumb.unset_parent();
+            }
+        }
+        info!(self.logger,"Selecting breadcrumb #{index}");
+    }
+
+    fn push_breadcrumb(&self, name:impl Str) {
+        let breadcrumb       = Breadcrumb::new(&self.scene,name);
+        let network          = &breadcrumb.frp.network;
+        let breadcrumb_index = self.breadcrumbs.borrow().len();
+        let model            = self.clone_ref();
+
+        frp::extend! { network
+            eval_ breadcrumb.frp.outputs.selected(model.select_breadcrumb(breadcrumb_index));
+        }
+
+        info!(self.logger,"Pushing {breadcrumb.name} breadcrumb");
+        breadcrumb.set_position(Vector3::new(self.width(),0.0,0.0));
+        self.add_child(&breadcrumb);
+        self.breadcrumbs.borrow_mut().push(breadcrumb);
+    }
+
+    fn pop_breadcrumb(&self) {
+        info!(self.logger, "Popping breadcrumb view.");
+        self.breadcrumbs.borrow_mut().pop().map(|breadcrumb| breadcrumb.unset_parent());
+    }
+}
+
+impl display::Object for BreadcrumbsModel {
+    fn display_object(&self) -> &display::object::Instance {
+        &self.display_object
+    }
+}
+
+
+
+// ===================
+// === Breadcrumbs ===
+// ===================
+
+/// The project name's view used for visualizing the project name and renaming it.
+#[derive(Debug,Clone,CloneRef,Shrinkwrap)]
+#[allow(missing_docs)]
+pub struct Breadcrumbs {
+    #[shrinkwrap(main_field)]
+    model   : Rc<BreadcrumbsModel>,
+    pub frp : Frp
+}
+
+impl Breadcrumbs {
+    /// Create a new ProjectName view.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S) -> Self {
```

Comment:
Why not use `Borrow<Scene>` rather than `Into<&Scene>?` That gives blanket implementations working both with references and values.